db/db_iter.cc

```diff
@@ -1234,7 +1245,7 @@ bool DBIter::FindUserKeyBeforeSavedKey() {
       PERF_COUNTER_ADD(internal_key_skipped_count, 1);
     }
 
-    if (num_skipped >= max_skip_ && CanReseekToSkip()) {
```

Comment:
Is not this case also vulnerable to infinite loop? Here we are doing Prev as well.

Reply:
The problem only occurs if you try to seek to max_visible_seq. Recall that reseek optimization applies in two cases 1. reseek to visible version of current key 2. reseek to next key. Write unprepared makes case 1 harder, but case 2 should "just work".

In this case, this is the seek we're doing:

```
last_key.SetInternalKey(ParsedInternalKey(saved_key_.GetUserKey(), kMaxSequenceNumber, kValueTypeForSeek));
```

We're simply seeking to the `kMaxSequenceNumber`, so we're guaranteed that the next key is they key we want.